### PR TITLE
Pull upstream

### PR DESCRIPTION
**Pull request by dszmigielski**
## Why

Keep in sync with upstraem.

## What

UpSteam Sync for S3.

## Tests

CI passed.

### CODE REVIEW

**Comment by Kielek** [comment]
File changes tab is unusable on my browser, some points to be clarified based on external tool review:
1. ~\docs\development\AutomaticInstrumentation.md - to be removed?
2. ~\tracer\src\Datadog.Trace\ClrProfiler\AutoInstrumentation\AWS\SQS\ContextPropagation.cs - SqsKey - _datadog - nor relate to the code changes

3. Compare ~\tracer\test\snapshots\AspNetCoreIisMvc21Tests.OutOfProcess.NoFF.__path=__statusCode=200.verified.txt to ~\tracer\test\snapshots\AspNetCoreIisMvc21Tests.OutOfProcess.WithFF.__path=__statusCode=200.verified.txt
WithFF vs NoFF - it looks strange, especially if we look into ~\tracer\test\Datadog.Trace.ClrProfiler.IntegrationTests\AspNetCore\AspNetCoreIisMvcTestBase.cs (_enableRouteTemplateResourceNames)
Similar things for other tests related to WithFF and NoFF. Please double check that changes are intentional.

4. Logger - direct submissions - do you have some short description of the feature? Should we support it?

5. A lot of new evn variables related to Logger - direct submissions - internal documentation should be updated.

File: tracer/src/Datadog.Trace/Configuration/TracerSettings.cs [hunk 1]
```diff
@@ -189,6 +190,7 @@ public TracerSettings(IConfigurationSource source)
             // If you change this, change environment_variables.h too
             ThreadSamplingEnabled = source?.GetBool(ConfigurationKeys.ThreadSampling.Enabled) ?? false;
             ThreadSamplingPeriod = GetThreadSamplingPeriod(source);
+            LogSubmissionSettings = new DirectLogSubmissionSettings(source);
```
Comment (pellared) [inline tracer/src/Datadog.Trace/Configuration/TracerSettings.cs:193]: please add a blank line before as the comment `// If you change this, change environment_variables.h too` does not applies for this line

File: docs/internal/internal-config.md [hunk 2]
```diff
@@ -13,6 +13,15 @@ These settings should be never used by the users.
 | `SIGNALFX_CONVENTION` | Sets the semantic and trace id conventions for the tracer. Available values are: `Datadog` (64bit trace id), `OpenTelemetry` (128 bit trace id). | `OpenTelemetry` |
 | `SIGNALFX_DUMP_ILREWRITE_ENABLED` | Allows the profiler to dump the IL original code and modification to the log. | `false` |
 | `SIGNALFX_EXPORTER` | The exporter to be used. The Tracer uses it to encode and dispatch traces. Available values are: `DatadogAgent`, `Zipkin`. | `Zipkin` |
+| `SIGNALFX_LOGS_DIRECT_SUBMISSION_INTEGRATIONS` | Configuration key for a list of direct log submission integrations to enable. Only selected integrations are enabled for direct log submission. Supports multiple values separated with semi-colons. | `` |
```
Comment (pellared) [inline docs/internal/internal-config.md:16]: As far as I understand the code the functionality is enabled ONLY if this option is set. Can we add supported values? I think these are the accptable values: `ILogger,Serilog,Log4Net,NLog`

**Comment by pellared** [comment]
What was wrong with `docs/development/AutomaticInstrumentation.md`? @Kielek why you wanted it to be removed?

File: tracer/test/snapshots/Security.AspNetMvc5.Integrated.enableSecurity=True.blockingEnabled=True.__url=_Health_wp-config.verified.txt [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+      http.url: http://localhost:00000/Health/wp-config,
+      language: dotnet,
+      net.peer.ip: ::1,
+      network.client.ip: ::1,
+      runtime-id: Guid_1,
+      signalfx.tracing.library: dotnet-tracing,
+      signalfx.tracing.version: x.y.z,
+      span.kind: server,
+      _dd.appsec.json: {"triggers":[{"rule":{"id":"nfd-000-001","name":"Detect common directory discovery scans","tags":{"type":"security_scanner","category":"attack_attempt"}},"rule_matches":[{"operator":"match_regex","operator_value":"^404$","parameters":[{"address":"server.response.status","key_path":[],"value":"404","highlight":["404"]}]},{"operator":"phrase_match","operator_value":"","parameters":[{"address":"server.request.uri.raw","key_path":[],"value":"http://localhost:00000/Health/wp-config","highlight":["/wp-config"]}]}]}]},
+      _dd.origin: appsec,
```
Comment (pellared) [inline tracer/test/snapshots/Security.AspNetMvc5.Integrated.enableSecurity=True.blockingEnabled=True.__url=_Health_wp-config.verified.txt]: we should check this

File: tracer/test/snapshots/Security.AspNetMvc5.Integrated.enableSecurity=True.blockingEnabled=True.__url=_Health_wp-config.verified.txt [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+      _dd.runtime_family: dotnet
```
Comment (pellared) [inline tracer/test/snapshots/Security.AspNetMvc5.Integrated.enableSecurity=True.blockingEnabled=True.__url=_Health_wp-config.verified.txt]: we should check this

**Comment by pellared** [comment]
It looks like the conflict resolution when merging https://github.com/DataDog/dd-trace-dotnet/pull/2178 was done incorrectly and most probably `DDSpanContextPropagator` does not work correctly. However, maybe we can simply remove it?

**Comment by pellared** [comment]
`SIGNALFX_TRACE_AZURE_FUNCTIONS_ENABLED` default is now `true` per https://github.com/DataDog/dd-trace-dotnet/pull/2326. 

1. Is it a problem for us? _I guess not, may be worth testing it somehow?_
2. At least our internal docs should be updated
2.

**Review by pellared — APPROVED** [review]
I have not found any blocke - approving.
However, some follow-up is needed.

**Comment by dszmigielski** [comment]
> I have not found any blocker - approving. However, some follow-up is needed.

I will apply as many of these as I can before merging it today

**Pull request merged**